### PR TITLE
Add support for persisting images/video in multiplayer

### PR DESCRIPTION
**Pull request by milo-**
Made a start on the multiplayer support for the image/video.

- [x] Store assets in Liveblocks 
- [x] Remove assets from asset table when deleting image/video shapes

### CODE REVIEW

File: examples/tldraw-example/src/multiplayer-with-images/multiplayer.tsx [hunk 1]
```diff
@@ -43,7 +43,6 @@ function Editor({ roomId }: { roomId: string }) {
          * without provifing a storage bucket based solution will cause
          * massive base64 string to be written to the liveblocks room.
          */
-        disableAssets={true}
```
Comment (FarazzShaikh) [inline examples/tldraw-example/src/multiplayer-with-images/multiplayer.tsx:46]: As mentioned in the comment, enabling assets without providing a storage bucket based encoding will cause massive base64 string to be written to Liveblocks. In my experimentation, love blocks can’t do large amount of data and will refuse to write videos (and very large images) as base64 outright.

**Comment by FarazzShaikh** [comment]
In my opinion, we should incorporate this as i have been trying to get multiplayer working and fails because the current multiplayer hook doesn’t share the asset table. This is greatly helpful! 🎉

File: examples/tldraw-example/src/multiplayer-with-images/multiplayer.tsx (hunk 1 above)
Comment (milo-) [reply examples/tldraw-example/src/multiplayer-with-images/multiplayer.tsx:46]: 👍 I've kept this on a hidden route (#/multiplayer-with-images) and using small png's to work around this limitation.

Will leave it up to @steveruizok to do the hosted image implementation :)

**Comment by milo-** [comment]
Thanks @FarazzShaikh! I've pushed some more commits to remove unused assets from the asset table (with undo/redo support).

I'm just working on tests now, and then I think this should be good to review :)

**Comment by FarazzShaikh** [comment]
Looking great! Can’t wait to have this in production!

**Comment by steveruizok** [comment]
Nice! Let me check this later today.

**Comment by steveruizok** [comment]
My initial thought about assets is that they might not have to be stored in Liveblocks, but rather in a generic bucket designed for this sort of static content (and perhaps with better options for avoiding abuse). Any thoughts?

**Comment by milo-** [comment]
I think the asset data will need to be stored in a bucket rather than liveblocks - I'm using the solution @FarazzShaikh has included for firebase in my own project and it's working great. 

The assets object that is then being stored in liveblocks is just metadata + src pointing to your bucket.

**Comment by FarazzShaikh** [comment]
Yes my original multiplayer solution was to store the actual assets in a bucket and have references to them in the document. With the asset table, the table itself (map of IDs to bucket URLs) can be stored as a part of the document in Liveblocks.

Each project can have its own bucket (or folder) this way all clients will write to and read from one bucket. 

To prevent abuse in the hosted instance of TLDraw, we can rate limit the number of assets pushed and also limit the size of an asset. Furthermore, if costs need to be kept to a minimum, we can limit the number of assets in a project. 

In the API, we can leave abuse prevention up to the implementation and provide a warning like the one I have written for enabling assets in multiplayer.

**Comment by fabianmu** [comment]
Yeah, having the assets in a bucket and reference them would be best and a great solution 👍.
Such a bucket URL, Credentials if needed and rate-limits could be in a configuration/env to easy adjust needs.

**Comment by steveruizok** [comment]
Sounds good to me! I'll take a look this week (ie when the nurseries are open again 👨‍👧).

**Comment by steveruizok** [comment]
Ok, I've set up an s3 bucket—two actually, one for dev and one for prod. Things are seeming to work correctly: files are uploaded and the src is saved in `assets` and propagated across LiveBlocks. I'll set the dev bucket permissions to allow uploads from the preview URL so we can test it out.

Kapture 2022-01-10 at 11 08 26

**Comment by FarazzShaikh** [comment]
Looking good! 

Maybe we can port the logic of the cursor interpolation to assets as they seem to suffer from lag but that’s for another time haha

**Comment by steveruizok** [comment]
Yeah all shapes update only on persist, so those temporary positions are lost. Working on a fix for that but interpolation could work too!

**Comment by milo-** [comment]
Nice one Steve, looking good! I do think I've spotted a small issue with onAssetDelete:

1. The ID provided to onAssetCreate/onAssetDelete is a shapeId. I think this should be an assetId so we have a common ID for creating/deleting shapes.
2. If an asset is shared between multiple shapes, the original asset would be deleted by the onAssetDelete callback:
https://github.com/tldraw/tldraw/blob/9cb96c2cf4ce4cca389c6421f5ae02afac16ac6f/packages/tldraw/src/state/TldrawApp.ts#L2570

**Comment by steveruizok** [comment]
I'm not sure we want to delete assets yet, because that would mean re-uploading them if a user pressed Undo after deleting the last shape that used the asset. We might mark assets for deletion and clean them up later?

**Comment by steveruizok** [comment]
I think this is good to go.

**Comment by steveruizok** [comment]
Nope, bug when deleting assets.

File: packages/tldraw/src/state/commands/shared/removeShapesFromPage.ts [hunk 2]
```diff
@@ -1,5 +1,14 @@
 import { TLDR } from '~state/TLDR'
-import type { ArrowShape, TDSnapshot, GroupShape, PagePartial } from '~types'
+import type { ArrowShape, GroupShape, PagePartial, TDPage, TDSnapshot } from '~types'
+
+const doesAssetExistInAnotherShape = (page: TDPage, deletedShapes: string[], assetId: string) =>
```
Comment (steveruizok) [inline packages/tldraw/src/state/commands/shared/removeShapesFromPage.ts:4]: Assets are shared between pages! I'm going to update this logic with a broader sweep at the end of the function.

**Comment by steveruizok** [comment]
I truly hope this feature doesn't give me a ten thousand dollar AWS bill.

**Comment by fabianmu** [comment]
I think there's a policy with which you can auto-delete all objects after x-days.
Not ideal for mid/long-term usage but as long as everything on tldraw.com is open and free most people will upload all sorts of images just to try and those will eat your £££. To make this feasible some sort of payment is needed.

**Comment by steveruizok** [comment]
Yeah, I may need to make images / uploads sponsors-only or else part of an eventual paid plan. It would be a while before the storage gets problematic though.

File: packages/tldraw/src/state/commands/shared/removeShapesFromPage.ts (hunk 2 above)
Comment (milo-) [reply packages/tldraw/src/state/commands/shared/removeShapesFromPage.ts]: 🙌 Thanks @steveruizok

**Comment by steveruizok** [comment]
Merged! The only thing that seems to be broken is getting the size from an SVG. I'll create a separate ticket for that.

Thanks again @milo- and @FarazzShaikh!

**Comment by milo-** [comment]
Woohoo!! Thanks @steveruizok and @FarazzShaikh 🎉

**Comment by FarazzShaikh** [comment]
Building good software one PR at a time 😄

**Pull request merged**